### PR TITLE
Feat custom domain

### PR DESCRIPTION
**Pull request by eddiejaoude**
## Fixes Issue

This feature, will allow the user to set a custom domain for their Profile, for example with the custom domain`biodrop.local:3000` (for local testing add to host file `/etc/hosts`) will display the web page from `localhost:3000/eddiejaoude`.

This is achieved by using NextJS middleware which will check Profiles in the DB for any match and return the profile on the custom domain without the appended `/eddiejaoude`

~~note: currently the `rewrite` rule (I left an inline comment) is super slow, commenting this out it is fine~~ local issue only

- [x] display profile on custom domain
- [x] form for user to add custom domain
- [x] test on Preview - manually adding custom domain
- [x] add custom domain to Vercel when adding domain to profile via premium form
   - [x] add / remove
   - [x] direct to project
   - [x] ssl
- [x] documentation for user to add custom domain + update their DNS
- [x] deploy to preview
   - [ ] ~~test domain~~
   - [ ] ~~test sub domain~~
- [x] deploy to prod
   - [ ] manage custom domain delete button (clear form field and submit)
   - [x] ~~middleware catch all and redirect on path for custom domain (for example: `custom-domain/search` -> `biodrop.io/search`)~~ hide header/footer
   - [ ] ~~search does not work due to CORS issue - not sure why this is effect by the custom domain in the middleware~~ because we used `promote preview`
   - [x] display domain status from vercel on BioDrop
   - [ ] sub domain error https://github.com/EddieHubCommunity/BioDrop/issues/9737

## Check List (Check all the applicable boxes) 

- [ ] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] This PR does not contain plagiarized content.
- [ ] The title of my pull request is a short description of the requested changes.

## Screenshots

Screenshot 2023-11-06 at 18 40 00

### CODE REVIEW

File: pages/api/profiles/[username]/[domain].js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  } catch (e) {
+    logger.error(e, `error finding profile for domain ${domain}`);
+    return {};
+  }
+
+  if (profile.length > 0) {
+    profile = profile[0];
+    return {
+      username: profile.username,
+      settings: { domain: profile.settings?.domain.replace("|", ".") },
```
[the remark this answers is not in the document]
Comment (eddiejaoude) [reply pages/api/profiles/[username]/[domain].js:55]: Good spot 👍

File: middleware.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  const profile = await res.json();
+
+  if (
+    profile.username &&
+    profile.user.type === "premium" &&
+    profile.settings?.domain &&
+    hostname === profile.settings.domain
+  ) {
+    // if match found rewrite to custom domain and display profile page
+    return NextResponse.rewrite(
```
Comment (eddiejaoude) [inline middleware.js:32]: This rewrite line is very slow, if I comment it out, it all runs fine

File: middleware.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      new URL(`/${profile.username}`, `http://${profile.settings.domain}`), // TODO: https
+    );
```
Comment (eddiejaoude) [inline middleware.js:34]: This rewrite line is very slow, if I comment it out, it all runs fine

Comment (eddiejaoude) [reply middleware.js]: It could be because of local development environment, I will deploy to preview on vercel and test it out

Comment (eddiejaoude) [reply middleware.js]: On Vercel Preview this is not an issue

**Comment by eddiejaoude** [comment]
Screenshot for docs

Screenshot 2023-11-04 at 16 04 42

**Comment by eddiejaoude** [comment]
Domain issue...

Screenshot 2023-11-04 at 17 01 19

**Comment by xkrishguptaa** [comment]
No Eddie that is not an issue @eddiejaoude; This also happens with hashnode: https://support.hashnode.com/en/articles/5755351-how-to-use-my-domain-on-both-vercel-and-hashnode

We have to display them this table like hashnode does; I believe

**Comment by eddiejaoude** [comment]
> No Eddie that is not an issue eddiejaoude; This also happens with hashnode: https://support.hashnode.com/en/articles/5755351-how-to-use-my-domain-on-both-vercel-and-hashnode
> 
> We have to display them this table like hashnode does; I believe

Ok, thank you 👍  great example 💯

**Comment by eddiejaoude** [comment]
Documentation screenshots

Screenshot 2023-11-06 at 11 27 03

Screenshot 2023-11-06 at 11 31 32

Screenshot 2023-11-06 at 11 37 45

**Pull request merged**